### PR TITLE
plutus-ir: add an optimization pass to merge let bindings

### PR DESCRIPTION
**Pull request by michaelpj**
This merges adjacent independent let-bindings, which can reduce the
nesting level of the program quite substantially. This is mostly only
important for human readability, but that's still useful.

Why not do this when we generate the let bindings in the first place?
Well, we need the dependency information to know when it's safe,
which we'd have to compute from the (related) information we have 
at that time, so it's not clear that it saves much work. Plus, this is
more general, and running it after eliminating dead lets makes it 
more effective.

### CODE REVIEW

File: plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+{- Note [Merging lets]
+We can merge independent, non-recursive, let bindings together into a single let binding, which is
+nice since it reduces the nesting of the program.
+
+We can *nearly* do this by just looking at the dependency graph and constructing level sets. However,
+nodes in the same level set may not be mergeable if their let bindings are not adjacent.
+
+So as it stands we need to traverse the term manually, performing local merging.
+-}
+mergeTerm
```
Comment (effectfully) [inline plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: This is not shown as a Haddock comment, right? I think, it's useful and small enough to be a Haddock comment.

File: plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+                Logically this should be the *reachable* set. However, given that we're already topologically
+                ordered, the only way there can be a transitive dependency between two adjacent nodes is if
+                there is a direct dependency. To see this, suppose that there were a transitive dependency
+                A -> B -> C, but A and C are adjacent. This is impossible, since B must come between A and C
+                in a topological ordering.
+                -}
+                mapMaybe Deps.nodeUnique $ Set.toList $ T.preSet (Deps.Variable u) deps
+
+        let (mergeableBindings, inner) = extractMergeable dependents t'
```
Comment (effectfully) [inline plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: Do I understand it right that

```
let x = 1
in let y = x
   in ...
```

does not get merged to

```
let x = 1
    y = x
in ...
```

?

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: Correct - that would require it to be a recursive let binding, which is more expensive to compute, so we don't do that.

Perhaps you're expecting let bindings to also be scoped top-to-bottom? I opted not to do that - if you want mutual scoping between bindings you need a recursive binding.

**Comment by michaelpj** [comment]
Hm, looks like I have a correctness bug.

File: plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs (hunk 2 above)
Comment (effectfully) [reply plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: > Perhaps you're expecting let bindings to also be scoped top-to-bottom?

Yep.

> I opted not to do that

I'm curious why.

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: Pros:
- Order-independence. 
    - You can always reorder bindings without changing the meaning.
- Precedent. 
    - At least: Ocaml, Nix.
- Simplicity. 
    - We already have a way of doing nested scopes with nested let bindings, no need to add it another way.

Cons:
- Can't merge as many let bindings. 
    - On the other hand, we already get blocked the moment we see a recursive let.
- More annoying to write/read.
    - It's still an IR so I don't care as much about that.

If I were writing a user-facing language I'd probably do something else, but for PIR I think this way is better.

Comment (effectfully) [reply plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: Thanks. I there a note in the code about that? If not, I think it's worth adding these pros and cons.

Comment (michaelpj) [reply plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs]: Perhaps the original design doc is the best place for this.

**Comment by michaelpj** [comment]
For future reference: I am planning to come back to this, but it has a mysterious bug that I haven't been able to reproduce except it seems to break the contract used by the playground and nothing else, which is worrying.

**Pull request closed without merging**